algebra-core/src/curves/cuda/scalar_mul/cpu_gpu_macros.rs

```diff
@@ -0,0 +1,246 @@
+
+                let arc_mutex = P::scalar_mul_static_profiler();
+                let mut profile_data = arc_mutex.lock().unwrap();
+                let mut proportions: Vec<f64> = profile_data.0.clone();
+
+                // If the program has just been initialised, we must check for the existence of existing
+                // cached profile data. If it does not exist, we create a new file
+                if proportions.is_empty() {
+                    let _ = std::fs::read_to_string(&dir.join("profile_data.txt"))
+                        .and_then(|s| { let res = serde_json::from_str(&s)?; Ok(res) })
+                        .and_then(|cached_data| {
+                            *profile_data = cached_data;
+                            proportions = profile_data.0.clone();
+                            Ok(())
+                        }
+                    );
+                }
+
+                if proportions.is_empty() {
+                    // By default we split the work evenly between devices and host
+                    proportions = vec![1.0 / (n_devices as f64 + 1.0); n_devices];
+                }
+
+                assert_eq!(proportions.len(), n_devices);
+                // Allocate the number of elements in the job to each device/host
+                let n_gpus = proportions.iter().map(|r| (r * n as f64).round() as usize).collect::<Vec<_>>();
+                let n_cpu = n - n_gpus.iter().sum::<usize>();
+
+                // Create storage for buffers and contexts for variable number of devices
+                let mut bases_split = Vec::with_capacity(n_devices);
+                let mut tables = Vec::with_capacity(n_devices);
+                let mut exps = Vec::with_capacity(n_devices);
+                let mut ctxs = Vec::with_capacity(n_devices);
+                let (mut time_cpu, mut times_gpu) = (0, vec![0; n_devices]);
+
+                // Split data and generate tables and u8 scalar encoding in device memory
+                for (i, &num) in n_gpus.iter().enumerate() {
+                    let device = Device::nth(i).unwrap();
+                    let ctx = device.create_context();
+
+                    let (lower, upper) = res_ref.split_at_mut(num);
+                    res_ref = upper;
+                    let lower_exps = &exps_h_ref[..num];
+                    exps_h_ref = &exps_h_ref[num..];
+
+                    let mut table = DeviceMemory::<Self>::zeros(&ctx, num * Self::table_size());
+                    let mut exp = DeviceMemory::<u8>::zeros(&ctx, num * Self::num_u8());
+
+                    Self::generate_tables_and_recoding(lower, &mut table[..], lower_exps, &mut exp[..]);
+
+                    ctxs.push((device, ctx));
+                    bases_split.push(lower);
+                    tables.push(table);
+                    exps.push(exp);
+                };
+
+                println!("Split statically and allocated device: {}us", now.elapsed().as_micros());
```

Comment:
This shouldn't appear in a normal run. Either change this to a debug print with logging or remove it.